portfolio/src/main/java/com/google/sps/data/HighestPriorityFirstScheduler.java

```diff
@@ -0,0 +1,114 @@
+      Comparator.comparing(ScheduledTask::getStartTime);
+
+  /** This method schedules tasks from highest to lowest priority and */
+  public Collection<ScheduledTask> schedule(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      Instant workHoursStartTime,
+      Instant workHoursEndTime) {
+
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    List<Task> tasksList = new ArrayList<Task>(tasks);
+
+    TaskQueue taskQueue = new TaskQueue(tasksList, getSchedulingAlgorithmType());
+
+    CalendarEventsGroup calendarEventsGroup =
+        new CalendarEventsGroup(eventsList, workHoursStartTime, workHoursEndTime);
+
+    List<TimeRange> availableTimes = calendarEventsGroup.getFreeTimeRanges();
+    TimeRangeGroup availableTimesGroup = new ArrayListTimeRangeGroup(availableTimes);
+    List<TimeRange> currentAvailableTimes = availableTimeRangesList(availableTimesGroup);
+
+    List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
+
+    int availableTimesIndex = 0;
+
+    Instant currentScheduleTime = workHoursStartTime;
+
+    while (availableTimesIndex < currentAvailableTimes.size() && !taskQueue.isEmpty()) {
+      TimeRange availableTimeRange = currentAvailableTimes.get(availableTimesIndex);
+      Task task = taskQueue.peek();
+
+      if (availableTimeRange.start().isAfter(currentScheduleTime)) {
+        currentScheduleTime = availableTimeRange.start();
+      }
+
+      if (!currentScheduleTime
+          .plusSeconds(task.getDuration().getSeconds())
+          .isAfter(availableTimeRange.end())) {
+        ScheduledTask scheduledTask =
+            new ScheduledTask(task, currentScheduleTime, Optional.of(true));
+        scheduledTasks.add(scheduledTask);
+
+        // Delete the TimeRange that has been scheduled over so that different priority tasks
+        // won't be scheduled over the same time.
+        availableTimesGroup.deleteTimeRange(
+            TimeRange.fromStartEnd(
+                scheduledTask.getStartTime(),
+                scheduledTask
+                    .getStartTime()
+                    .plusSeconds(scheduledTask.getTask().getDuration().getSeconds())));
+        // After deleting a TimeRange, we must renew the list.
+        currentAvailableTimes = availableTimeRangesList(availableTimesGroup);
+
+        // Push the currentScheduleTime back by the amount that was scheduled.
+        currentScheduleTime = currentScheduleTime.plusSeconds(task.getDuration().getSeconds());
+        taskQueue.remove();
+
+        // If the next task's priority is different from the task that was just scheduled,
+        // then reset the availableTimesIndex, currentScheduleTime, and make the loop
+        // grab a TimeRange. We go back to the first availableTimeRange in order to schedule
+        // more tasks towards the beginning of the availableTimes.
+        if (!taskQueue.isEmpty()
```

Comment:
Maybe it's cleaner to pull this condition into a helper method called `isNextTaskSamePriority` ?

Reply:
Done.